cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetTableMigrator.java

```diff
@@ -0,0 +1,175 @@
+import org.apache.hadoop.hbase.client.Put;
+import org.apache.hadoop.hbase.client.Result;
+import org.apache.hadoop.hbase.client.ResultScanner;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Map;
+import java.util.NavigableMap;
+
+/**
+ * Migrates the tables of a PartitionedFileSet.
+ */
+public class PartitionedFileSetTableMigrator {
+  private static final Logger LOG = LoggerFactory.getLogger(PartitionedFileSetTableMigrator.class);
+  private static final long MILLION = 1000 * 1000;
+  protected final HBaseTableUtil tableUtil;
+  protected final Configuration conf;
+  private final DatasetFramework dsFramework;
+
+  @Inject
+  protected PartitionedFileSetTableMigrator(HBaseTableUtil tableUtil, Configuration conf,
+                                            DatasetFramework dsFramework) {
+    this.tableUtil = tableUtil;
+    this.conf = conf;
+    this.dsFramework = dsFramework;
+  }
+
+  /**
+   * Given a specification of a PartitionedFileSet, creates an IndexedTable for the new embedded partitions Dataset
+   * and migrates the data from the old partitions table to the new partitions table.
+   *
+   * @param namespaceId the namespace that contains the PartitionedFileSet
+   * @param dsSpec the specification of the PartitionedFileSet that needs migrating
+   * @throws Exception
+   */
+  public void upgrade(Id.Namespace namespaceId, DatasetSpecification dsSpec) throws Exception {
+    DatasetSpecification newPartitionsSpec = dsSpec.getSpecification(PartitionedFileSetDefinition.PARTITION_TABLE_NAME);
+    TableId oldPartitionsTableId = TableId.from(namespaceId, newPartitionsSpec.getName());
+    TableId indexTableId = TableId.from(namespaceId, newPartitionsSpec.getName() + ".i");
+    TableId dataTableId = TableId.from(namespaceId, newPartitionsSpec.getName() + ".d");
+
+
+    byte[] columnFamily = HBaseTableAdmin.getColumnFamily(newPartitionsSpec);
+
+    HBaseAdmin hBaseAdmin = new HBaseAdmin(conf);
+    if (!tableUtil.tableExists(hBaseAdmin, oldPartitionsTableId)) {
+      LOG.info("Old Partitions table does not exist: {}. Nothing to migrate from.", oldPartitionsTableId);
+      return;
+    }
+
+    DatasetsUtil.createIfNotExists(dsFramework, Id.DatasetInstance.from(namespaceId, newPartitionsSpec.getName()),
+                                   IndexedTable.class.getName(),
+                                   DatasetProperties.builder().addAll(newPartitionsSpec.getProperties()).build());
+
+    HTable oldTable = tableUtil.createHTable(conf, oldPartitionsTableId);
+    HTable newIndexTable = tableUtil.createHTable(conf, indexTableId);
+    HTable newDataTable = tableUtil.createHTable(conf, dataTableId);
+
+    LOG.info("Starting upgrade for table {}", Bytes.toString(oldTable.getTableName()));
+    try {
+      ScanBuilder scan = tableUtil.buildScan();
+      scan.setTimeRange(0, HConstants.LATEST_TIMESTAMP);
+      scan.setMaxVersions(1); // we only need to see one version of each row
```

Comment:
Not true. What if a partition was created by a MR job, and that job then failed? The transaction would be invalidated (and not rolled back). That is, the write remains, but its write version is now in the invalid list maintained by the tx manager. So, you need to talk to tx manager, get a transaction. Count the number of excludes in that transaction, and that is the max version that you need. Then, while reading, you need to filter out excluded versions.